### PR TITLE
📖  Sync cluster-api-state-metrics proposal to match implementation state.

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Syncs the cluster-api-state-metrics proposal to match implementation state.
Implementation was mostly done in #7095

This also removes the `_labels` metric references. There are kube-state-metrics specific flags regarding these metric which won't affect custom resource configuration (allow/deny-list). Exposing all metrics would lead increase the cardinality without being worth it. Use-case specific configuration should be added by the user if required.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #6458

### CODE REVIEW

File: docs/proposals/20220411-cluster-api-state-metrics.md
Comment (sbueringer) [inline docs/proposals/20220411-cluster-api-state-metrics.md:118]: l.124 with the mention of "cluster-api-state-metrics" is confusing

File: docs/proposals/20220411-cluster-api-state-metrics.md
Comment (sbueringer) [inline docs/proposals/20220411-cluster-api-state-metrics.md:192]: That is a very interesting point. Although I'm not sure if it's worth building something to generate them (we need a source file somewhere for that)

Maybe writing them manually is good enough?

File: docs/proposals/20220411-cluster-api-state-metrics.md
Comment (sbueringer) [inline docs/proposals/20220411-cluster-api-state-metrics.md:228]: This seems outdated given that with "the spirit of the kube-state-metrics maintainers" we meant they don't want to support CRDs.

(but we can just drop this part of the sentence)

File: docs/proposals/20220411-cluster-api-state-metrics.md
Comment (sbueringer) [inline docs/proposals/20220411-cluster-api-state-metrics.md:376]: Is _labels something that could be interesting in the future?

Afaik they support it for node_labels and it can be very very useful

File: docs/proposals/20220411-cluster-api-state-metrics.md
Comment (chrischdi) [reply docs/proposals/20220411-cluster-api-state-metrics.md:192]: Could be also done by the configuration generator.

By knowing that we have a configuration for the api resource, it also knows that it requires an rbac rule to allow access to the resource.
The config includes the same information as needed for rbac in that case, except the verbs (which are always `get`,`list`,`watch`).

Comment (sbueringer) [reply docs/proposals/20220411-cluster-api-state-metrics.md:192]: Sounds good. Can we generalize the wording a bit. Currently it states `via kubebuilder annotations`.

File: docs/proposals/20220411-cluster-api-state-metrics.md
Comment (chrischdi) [reply docs/proposals/20220411-cluster-api-state-metrics.md:376]: A generic `_labels` metric which also re-uses the same flags as for `kube_*_labels` metrics is currently not possible.

However, it is possible to configure your custom `Info` metric which behaves similar but requires an explicit listing of which labels to expose.
To not risk high cardinality and due to the lack of support I removed it for now.

Don't know if we could highlight this in a better way. Maybe at future work?

Comment (sbueringer) [reply docs/proposals/20220411-cluster-api-state-metrics.md:379]: Ah good to know that it can be done via `Info`. Could be very useful to expose some well-known labels that we know are there.

I would say the current state is fine and we can eventually document it when we get to documenting "how to generate your metrics config". Let's just resolve for now

Comment (sbueringer) [reply docs/proposals/20220411-cluster-api-state-metrics.md:379]: But it's an interesting use case for the metric gen tool as it's not super straight-forward

File: docs/proposals/20220411-cluster-api-state-metrics.md
Comment (sbueringer) [inline docs/proposals/20220411-cluster-api-state-metrics.md:68]:
```suggestion
This solution is derived from [mercedes-benz/cluster-api-state-metrics] and all the merits for its inception goes to the team that created it: [@chrischdi](https://github.com/chrischdi), [@seanschneeweiss](https://github.com/seanschneeweiss), [@tobiasgiese](https://github.com/tobiasgiese).
```
nit

**Review by sbueringer — COMMENTED** [review]
last nit from my side. otherwise lgtm pending squash

File: docs/proposals/20220411-cluster-api-state-metrics.md
Comment (chrischdi) [reply docs/proposals/20220411-cluster-api-state-metrics.md:379]: I think its more a feature on ksm side / enable a `_labels` metric as for any `kube_*_labels` metric, but just for custom resource.

**Comment by sbueringer** [comment]
lgtm pending squash

**Comment by fabriziopandini** [comment]
/lgtm

it would be great if @tobiasgiese (co-author on this proposal) could take a quick look, otherwise, I will approve in a couple of days

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Comment by tobiasgiese** [comment]
> it would be great if @tobiasgiese (co-author on this proposal) could take a quick look, otherwise, I will approve in a couple of days

/lgtm
Thanks for this great rework of the design proposal 🙌🏻

**Pull request merged**
